add solver: succinct small progress measures

pazz opened the pull request:
Apologies for the hickups; somehow the github interface behaves strangely with this commit.
You're welcome to ignore this PR. I just wanted to keep it around so that I can find it.

---------------

This adds an implementation the "succint small progress measures"
algorithm proposed by Jurdzinski/Lazic in https://arxiv.org/abs/1702.05051.

The special comparisons for bitstrings and derived bounded adaptive
counters are done explicitly in ocaml, which is slow.

So far, this only derives a winning strategy for player 0
and sets the strategy for player 1 to a trivial bogus strategy.

oliverfriedmann commented:
Great, thank you! We'll pull it in as soon as it solves the player 1 strategy as well. Not sure if it helps, but you can use the included functions to 1) create a subgame from the player 1 winning region 2) create the "dual" game in which player 0 wins 3) recursively call your algorithm 4) transform everything back

oliverfriedmann commented:
Pulled it into this branch:

https://github.com/tcsprojects/pgsolver/tree/succinct-algorithm

Can you have a look at this game?

parity 10;
0 6 1 2,5,7 "0";
1 6 0 0,2,9 "1";
2 3 1 3,4,9 "2";
3 4 0 1,4 "3";
4 9 0 4,5 "4";
5 4 1 1,3 "5";
6 8 0 0,2,8 "6";
7 2 1 6,7 "7";
8 10 1 1,5,8 "8";
9 5 1 3,4,8 "9";

It is completely won by player 1 but the implementation claims it is completely won by player 0.

pazz commented:
I've just rebased my branch to tcsprojects:develop without testing. @oliverfriedmann: I've seen that you have changed the way solvers are registered. I hope my amended patch makes sense in this respect.

In other news, @aml119 claims to have fixed the bug you found.
His repo is here: https://github.com/aml119/pgsolver/tree/succinct-algorithm
and I have added his patch to the PR here as well.

oliverfriedmann commented:
Confirmed that is working now. Pulled into master. I took the liberty and added some auxiliary code to also compute the strategy for player 1 by calling the algorithm again on the dual sub game.

pazz commented:
that was quick, thanks :-)
